### ISSUE TITLE
Error: unused arguments

### ISSUE THREAD

**Issue by ashokpatowary** [GoekeLab/bambu-singlecell-spatial#6]
Hi @lingminhao 

I was trying to run bambu-sc in my test data and encountering the following error. Can you please give a look

`  Error in bambu(reads = samples, annotations = annotations, genome = "GRCm39.primary_assembly.genome.fa",  : 
    unused arguments (processByChromosome = as.logical("TRUE"), processByBam = as.logical("false"))
  Execution halted`

Thanks

**Comment by ashokpatowary** [comment]
Thanks @lingminhao . I did try with the new singularity package; however encounter the following error

`Running Bambu-v3.9.0
WARNING - If you change the number of cores (ncore) between Bambu runs and there is no progress please restart your R session to resolve the issue that originates from the XGboost package.
  --- Start generating read class files ---
  Error: BiocParallel errors
    1 remote errors, element index: 1
    0 unevaluated and other errors
    first remote error:
  Error: object 'true' not found
  Execution halted
  INFO:    Cleaning up image...`

**Comment by lingminhao** [comment]
Hi @ashokpatowary, I have fixed the bug regarding your errors. Let me know if that helps.

**Comment by ashokpatowary** [comment]
Thanks @lingminhao ; but this time encounter an additional error.

`Running Bambu-v3.9.0
  WARNING - If you change the number of cores (ncore) between Bambu runs and there is no progress please restart your R session to resolve the issue that originates from the XGboost package.
  --- Start generating read class files ---
  Error: BiocParallel errors
    1 remote errors, element index: 1
    0 unevaluated and other errors
    first remote error:
  Error in clipFunction(alignData = GenomicAlignments::cigar(alignmentInfo), : unused argument (alignData = GenomicAlignments::cigar(alignmentInfo))
  Execution halted
  INFO:    Cleaning up image...`

**Comment by lingminhao** [comment]
Hi @ashokpatowary , I think you will have to update your singularity image to the latest ones by pulling it again.

**Comment by ashokpatowary** [comment]
Thanks @lingminhao. This time it not showing the previous error; but I encounter a different error. Looks like its input  data related; can you please give your input

`prediction accuracy (CV) (higher for splice donor than splice acceptor)
  estimate: 12.7803086562292
  pValue: 0.0138140193757546
  AUC: 0.88
  prediction accuracy (CV) (higher for splice donor than splice acceptor)
  Error in fisher.test(table(predictions > 0.5, labels.train.cv.test)) : 
    'x' must have at least 2 rows and columns
  Calls: bambu ... FUN -> testSpliceSites -> fitXGBoostModel -> fisher.test
  Execution halted
  INFO:    Cleaning up image...
`

The code I am using 

`nextflow run bambu-singlecell-spatial/main.nf --bams Run28_sort.bam --genome reference/GRCm39.primary_assembly.genome.fa --annotation reference/gencode.vM34.annotation.gtf --ncore 36 --outdir output -with-singularity lingminhao/bambusc:beta1.2 --cleanReads TRUE --deduplicateUMIs TRUE --processByBam FALSE --processByChromosome TRUE`

**Comment by lingminhao** [comment]
Hi @ashokpatowary, I checked the error, it turns out that it occurs at parts where it was only used to provide the user with message about some statistics used in the model, and will not impact the output. I turned off this functionality, so you should be able to get rid over this error now. There's no need to pull new singularity image for this error.  

Having said so, it will be beneficial if you could provide the test data if possible, or the `.command.run` & `.command.log` files under the work directory for third process in nextflow: `bambu`, so that we can diagnose the bug in detail and fix it accordingly. But let's see if there are any further errors first.

**Comment by ashokpatowary** [comment]
Hi @lingminhao. Thanks for all your help. I got another error this time. I have uploaded all the files here: https://www.dropbox.com/scl/fo/6xybedt5u4dqjzj0xc0k4/AMIrTuRO1PesQUBzLDWtGoU?rlkey=v2lrxt8yc0zqbg9mknqlkq8r9&st=j0ajbihk&dl=0

`--- Start extending annotations ---
  Error in lm.fit(x, y, offset = offset, singular.ok = singular.ok, ...) : 
    0 (non-NA) cases
  Calls: bambu ... filterTranscriptsByAnnotation -> recommendNDR -> predict -> lm -> lm.fit
  Execution halted
  INFO:    Cleaning up image...
`

**Comment by lingminhao** [comment]
Hi @lingminhao , I did a check with your dataset, and realised that the `Rsamtools` package parsed the information from the BC and UG tag in the bam file incorrectly. This can happen if these BC & UG tag are added manually into the bam file, there could be errors during the process. 

![Image](https://github.com/user-attachments/assets/cf27f130-4b42-454d-8773-9df73b50d95b)

Nevertheless, we acknowledge this can be a potential issue in the future, so now we instead prioritise BC/UMI parsing from the read name (if satisfy BC_UMI#READNAME format) over the tags in the bam file. You may proceed without modifying anything in the bam file. 

PS: I also tried running your downsampled dataset, and realise it's insufficient to run with 64G RAM on my mobile machine (I am on a trip), mostly due to the high number of cells. You may have to run on a machine with higher RAM. 

Thanks!

**Comment by ashokpatowary** [comment]
Thanks @lingminhao 

Data is from non 10X processing; so I added the BC and UMI tag manually. Anyways, thanks for the help and yess this was very small dataset from large number of cells. But our final data will have over 2B reads.

Regards
Ashok

**Comment by lingminhao** [comment]
Hi @ashokpatowary, 

sorry I forgot to mention again that you'll need to update the singularity image again for this BC&UMI parsing. 

Yes, we haven’t been working much with datasets from non-10x platforms, those with over 60k cells, or 2 billion reads. So, your feedback here is very helpful to us!

Best regards, 
Min Hao

**Comment by ashokpatowary** [comment]
@lingminhao 

I think if we give raw bam file as input; there will be always memory error. But if we clean the input data bases on Knee; that can be avoided. Saying that I am encountering the following error. Any suggestion where it went wrong

`  Running Bambu-v3.9.0
  WARNING - If you change the number of cores (ncore) between Bambu runs and there is no progress please restart your R session to resolve the issue that originates from the XGboost package.
  --- Start isoform EM quantification ---
  Error in fread(clusters[[j]], header = FALSE, data.table = FALSE) : 
    input= must be a single character string containing a file name, a system command containing at least one space, a URL starting 'http[s]://', 'ftp[s]://' or 'file://', or, the input data itself containing at least one \n or \r
  Calls: bambu -> fread -> stopf -> raise_condition -> signal
  Execution halted
  INFO:    Cleaning up image...`

**Comment by ashokpatowary** [comment]
@lingminhao 

I think it's the "EM" which causing the error. I did try running with 10X data; same error. Saying that I have another concern. I used the sample sample to generate

1. 10X + PacBio
2. non-10X+ONT

For 10X+PacBio; I did analysis previously with a different tool; I did used lots of experimental validation about novel finding. To my surprise when I used that same data with bambu-single-cell less than 5% of the novel transcripts were called. I am not sure why this happening. can we used `opt.discovery` in this; that might be an issue? Additionally; since lots of reads which might be a novel transcripts are filtered out; while calculating gene count matrix; does these filtered reads considered or not? For example; one reads might be incomplete splice matched for thew transcript; but for the gene count that will be a true copy.

**Comment by lingminhao** [comment]
Hi @ashokpatowary,

Let me break your questions into bullet point so it's easier. 

- the bug is now fixed, remember to pull a new singularity image. 
- Can I confirm with you what you meant by knee? Is that the knee plot used to filter the cell barcodes? More info about this is very helpful for future improvement for dataset with large number of cells 
- For more novel transcript discovery, you can try to increase the threshold using the `NDR` argument. A higher `NDR` meant to include more novel transcript, with the risk of increased number of false discovery. `opt.discovery` is used when your annotation is incomplete for model training or that you can only use annotations from another species for novel transcript discovery
- Even though these reads are not sufficient to evident existence of novel transcript, they are still used for quantifications.

**Comment by ashokpatowary** [comment]
Hi @lingminhao 

Thanks for your response. Yes, I was talking about knee plot to remove background. I think with removal of the background it's efficiency increase. 

With NDR=1, I still get the following error; but its related to EM so not very imp for me at this moment

`Error in reducer$value.cache[[as.character(idx)]] <- values : 
    wrong args for environment subassignment
  Calls: bambu ... .bploop_impl -> .collect_result -> .reducer_add -> .reducer_add
  In addition: Warning message:
  In parallel::mccollect(wait = FALSE, timeout = 1) :
    1 parallel job did not deliver a result
  Execution halted`

I have few other queries 

1. If I gradually increase NDR from o.2, 0.4, ..1; I suddenly see a huge increase in number of transcript for NDR value 0.8 to 1 (~4x). I dint see this while using bulk samples. Is it expected? Does the transcript include "incomplete splice match" category. I prefer not a blanket filtering of this category.  There will be artifacts; but in my 2022 Science study I did validate a huge number of these (though not reported in the paper). I will prefer to filter using my own criteria if possible.
2.  There are multiple gtf in the output (extended_annotations_NDR1.gtf is not present). If I need to extract all the transcripts present in the data which one I should consider; or how can I filter it (similar to full-length>1 in bulk)
```
combined_extended_annotations_allTranscriptModels.gtf
combined_extended_annotations_novelTranscripts.gtf
combined_extended_annotations_extendedAnnotations.gtf
combined_extended_annotations_supportedTranscriptModels.gtf
```
Thanks

**Comment by lingminhao** [comment]
Hi @ashokpatowary, hope you are doing well. To answer your queries,

1. I did not obtain the error you mentioned when tested on our test dataset when NDR is set to 1 . Maybe this issue is data-specific. Feel free to let me know if you require the output related to EM in future, I am happy to debug it if you're able to provide a subsampled dataset from your side. 
2. You found a lot more novel transcript in the sc sample than the bulk sample. This is because in bambu-sc, we used a newer version of bambu (which is still not released actually), where we remoes all filter like subset transcript (incomplete splice match), or the gene proportions criteria when using NDR of 1.
3. Since you are using a NDR of 1, and there's no filtering, all putative novel transcript (indicated with "BambuTxxxx") are included, you may filter with your own filtering criteria. You can find these novel transcript in  

- `combined_extended_annotations_allTranscriptModels.gtf` that gives everything (NDR =1)
- `combined_extended_annotations_novelTranscripts.gtf` that gives only the output about the novel transcript  
- `combined_extended_annotations_extendedAnnotations.gtf` that gives the output based on the user-specified NDR
- combined_extended_annotations_supportedTranscriptModels.gtf that gives the output about the transcript that has full length support 

Best regards, 
Min Hao

**Comment by ashokpatowary** [comment]
Thanks @lingminhao 

We are generating additional data and will try it with ~2 Billion reads. I have closed this thread for now as your manuscript might be in submission process. I might reopen it in coming days.

Thanks
